daemon/rename.go

```diff
@@ -27,22 +38,53 @@ func (daemon *Daemon) ContainerRename(oldName, newName string) error {
 
 	container.Name = newName
 
-	undo := func() {
-		container.Name = oldName
-		daemon.reserveName(container.ID, oldName)
-		daemon.containerGraphDB.Delete(newName)
+	undo = append(undo,
+		func() {
+			container.Name = oldName
+			daemon.reserveName(container.ID, oldName)
+			daemon.containerGraphDB.Delete(newName)
+		})
+
+	if err = daemon.containerGraphDB.Delete(oldName); err != nil {
+		err = derr.ErrorCodeRenameDelete.WithArgs(oldName, err)
+		goto err_return
 	}
 
-	if err := daemon.containerGraphDB.Delete(oldName); err != nil {
-		undo()
-		return derr.ErrorCodeRenameDelete.WithArgs(oldName, err)
+	if err = container.toDisk(); err != nil {
+		goto err_return
 	}
 
-	if err := container.toDisk(); err != nil {
-		undo()
-		return err
+	if !container.Running {
+		container.logEvent("rename")
+		return nil
+	}
+
+	sid = container.NetworkSettings.SandboxID
+	sb, err = daemon.netController.SandboxByID(sid)
+	if err != nil {
+		undo = append(undo,
+			func() {
+				container.Name = oldName
+				if e := container.toDisk(); e != nil {
+					logrus.Errorf("%s: Failed in writing to Disk on rename failure: %v", container.ID, e)
+				}
+			})
+		goto err_return
+	}
+
+	err = sb.Rename(strings.TrimPrefix(container.Name, "/"))
+	if err != nil {
+		goto err_return
 	}
 
 	container.logEvent("rename")
-	return nil
+
+err_return:
+	if err != nil {
+		for _, f := range undo {
```

Comment:
Isn't "a stack of calls that execute when the function unwinds" the definition of `defer`? I'm not sure I see why the manual management of the `undo` array is needed here.